src/pardibaal/Federation.cpp

```diff
@@ -66,6 +66,38 @@ namespace pardibaal {
             zones.push_back(dbm);
     }
 
+    void Federation::add(const Federation& fed) {
+        for (const auto& dbm : fed)
+            this->add(dbm);
+    }
+
+    void Federation::subtract(const DBM &dbm) {
+#ifndef NEXCEPTIONS
+        if (!zones.empty()) {
+            if (dimension() != dbm.dimension())
+                throw base_error("ERROR: Subtracting dbm with dimension: ", dbm.dimension(),
+                                 " from a federation with dimension: ", dimension());
+        }
+#endif
+        auto fed = Federation();
+        for (auto z : zones) {
+            for (dim_t i = 0; i < dimension(); ++i) {
+                for (dim_t j = 0; j < dimension(); ++j) {
+                    if (z.at(i, j) > dbm.at(i, j)) {
+                        z.restrict(j, i, bound_t(-dbm.at(i, j).get_bound(), dbm.at(i, j).is_non_strict()));
+                        fed.add(z);
```

Comment:
Without checking here; what happens if Z is empty? and Z is the first zone added to the result? 
And wouldn't you need to first do a canonization to be able to check emptiness ?